R/build-home-index.R

```diff
@@ -39,17 +45,50 @@ data_home <- function(pkg = ".") {
 
 data_home_sidebar <- function(pkg = ".") {
   pkg <- as_pkgdown(pkg)
-  if (!is.null(pkg$meta$home$sidebar))
+  if (isFALSE(pkg$meta$home$sidebar))
     return(pkg$meta$home$sidebar)
 
+  sidebar_path <- file.path(pkg$src_path, pkg$meta$home$sidebar$html)
+
+  if (length(sidebar_path)) {
+
+    if (!file.exists(sidebar_path)) {
+      abort(sprintf("Can't find file %s.", pkg$meta$home$sidebar$html))
+    }
+    return(paste0(read_lines(sidebar_path), collapse = "\n"))
+  }
+
+  if (is.null(pkg$meta$home$sidebar$structure)) {
+    return(paste0(
+      data_home_sidebar_links(pkg),
+      data_home_sidebar_license(pkg),
+      data_home_sidebar_community(pkg),
+      data_home_sidebar_citation(pkg),
+      data_home_sidebar_authors(pkg),
+      collapse = "\n"
+    ))
+  }
+
   paste0(
-    data_home_sidebar_links(pkg),
-    data_home_sidebar_license(pkg),
-    data_home_sidebar_community(pkg),
-    data_home_sidebar_citation(pkg),
-    data_home_sidebar_authors(pkg),
+    purrr::map_chr(pkg$meta$home$sidebar$structure, data_home_element, pkg = pkg),
     collapse = "\n"
   )
+
+
+}
+
+data_home_element <- function(element, pkg) {
+
+  if (element %in% c("links", "license", "community", "citation", "authors")) {
+    return(do.call(paste0("data_home_sidebar_", element), list(pkg = pkg)))
+  }
+  component <- pkg$meta$home$sidebar$components[[element]]
+
+  if (is.null(component)) {
+    abort(sprintf("There is no component named %s in the components field", element))
```

Comment:
I am still unsure how to best write error messages about nested YAML fields.

Reply:
Maybe in an interactive session it'd be useful to even open the YAML file with the cursor at the right position... although that's a bit too much.

Reply:
I think we need to start with some convention to how we refer to nested yaml fields. Do we use `$`? Do we use `.`? And then probably make a little helper so we can write `field("a", "b", "c")` to generate "a.b.c (in _pkgdown.yml)" or similar.  (Maybe colouring the path to make it obvious that it's something special.

Reply:
I used "." because I thought it looked less busy but `home.sidebar.html` can confusingly look like a path.